codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -60,38 +60,31 @@ public Http2ConnectionHandler(Http2Connection connection, Http2FrameListener lis
     }
 
     public Http2ConnectionHandler(Http2Connection connection, Http2FrameReader frameReader,
-            Http2FrameWriter frameWriter, Http2FrameListener listener) {
-        this(DefaultHttp2ConnectionDecoder.newBuilder().connection(connection)
-                .frameReader(frameReader).listener(listener),
-             DefaultHttp2ConnectionEncoder.newBuilder().connection(connection)
-                .frameWriter(frameWriter));
+                                  Http2FrameWriter frameWriter, Http2FrameListener listener) {
+        checkNotNull(connection, "connection");
+        checkNotNull(frameReader, "frameReader");
+        checkNotNull(frameWriter, "frameWriter");
+        checkNotNull(listener, "listener");
+        encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
+        decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, frameReader, listener);
+        init();
     }
 
     /**
-     * Constructor for pre-configured encoder and decoder builders. Just sets the {@code this} as the
+     * Constructor for pre-configured encoder and decoder. Just sets the {@code this} as the
      * {@link Http2LifecycleManager} and builds them.
      */
-    public Http2ConnectionHandler(Http2ConnectionDecoder.Builder decoderBuilder,
-            Http2ConnectionEncoder.Builder encoderBuilder) {
-        checkNotNull(decoderBuilder, "decoderBuilder");
-        checkNotNull(encoderBuilder, "encoderBuilder");
-
-        if (encoderBuilder.lifecycleManager() != decoderBuilder.lifecycleManager()) {
-            throw new IllegalArgumentException("Encoder and Decoder must share a lifecycle manager");
-        } else if (encoderBuilder.lifecycleManager() == null) {
-            encoderBuilder.lifecycleManager(this);
-            decoderBuilder.lifecycleManager(this);
-        }
-
-        // Build the encoder.
-        encoder = checkNotNull(encoderBuilder.build(), "encoder");
+    public Http2ConnectionHandler(Http2ConnectionDecoder decoder,
+                                  Http2ConnectionEncoder encoder) {
+        this.decoder = checkNotNull(decoder, "decoder");
+        this.encoder = checkNotNull(encoder, "encoder");
+        init();
```

Comment:
I think we should avoid the escaping constructor issue.  We may be OK now but changes in the future, or extensions of our classes may experience half initialized object issues.  Static factory method is the first thing that comes to mind...can anyone think of anything more clever?